shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,254 @@
+//------------------------------------------------------------------------------
+
+#if defined(HAS_REFLECTIONS) && REFLECTION_MODE == REFLECTION_MODE_SCREEN_SPACE
+
+// Copied from depthUtils.fs
+highp float linearizeDepth(highp float depth) {
+    // Our far plane is at infinity, which causes a division by zero below, which in turn
+    // causes some issues on some GPU. We workaround it by replacing "infinity" by the closest
+    // value representable in  a 24 bit depth buffer.
+    const highp float preventDiv0 = 1.0 / 16777216.0;
+    highp mat4 p = getViewFromClipMatrix();
+    // this works with perspective and ortho projections, for a perspective projection
+    // this resolves to -near/depth, for an ortho projection this resolves to depth*(far - near) - far
+    return (depth * p[2].z + p[3].z) / max(depth * p[2].w + p[3].w, preventDiv0);
+}
+
+// Code adapted from McGuire and Mara, Efficient GPU Screen-Space Ray Tracing, Journal of Computer
+// Graphics Techniques, 2014
+//
+// Copyright (c) 2014, Morgan McGuire and Michael Mara
+// All rights reserved.
+//
+// This software is open source under the "BSD 2-clause license":
+//
+//    Redistribution and use in source and binary forms, with or without modification, are permitted
+//    provided that the following conditions are met:
+//
+//    1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+//    2. Redistributions in binary form must reproduce the above copyright notice, this list of
+//    conditions and the following disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//
+//    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+//    IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+//    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+//    CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+//    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+//    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+//    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+//    POSSIBILITY OF SUCH DAMAGE.
+
+void swap(inout highp float a, inout highp float b) {
+     highp float temp = a;
+     a = b;
+     b = temp;
+}
+
+highp float distanceSquared(highp vec2 a, highp vec2 b) {
+    a -= b;
+    return dot(a, a);
+}
+
+// Note: McGuire and Mara use the "cs" prefix to stand for "camera space", equivalent to Filament's
+// "view space". "cs" has been replaced with "vs" to avoid confusion.
+bool traceScreenSpaceRay(const highp vec3 vsOrigin, const highp vec3 vsDirection,
```

Comment:
We talked about  moving the raycast code into its own file so it can be used for other things eventually. Ok to do in a later PR.